client/app/types/channels/liveMonitoring.ts

```diff
@@ -0,0 +1,53 @@
+export interface MonitoringMonitorData {
+  maxIntervalMs: number;
+  offsetMs: number;
+  hasSEBHash: boolean;
+}
+
+export interface HeartbeatDetail {
+  isValidSEBHash: boolean;
+  stale: boolean;
+  userAgent: string;
+  ipAddress: string;
+  generatedAt: string;
+  sebHash?: string;
+}
+
+export interface SnapshotData {
+  sessionId: number;
+  status: 'expired' | 'listening' | 'stopped';
+  lastHeartbeatAt: string;
+  isValidSEBHash: boolean;
+  sebHash?: string;
+  userName?: string;
+  stale?: boolean;
+}
+
+export type SnapshotsData = Record<number, SnapshotData>;
+
+export interface Snapshot extends SnapshotData {
+  selected?: boolean;
+  recentHeartbeats?: HeartbeatDetail[];
+  hidden?: boolean;
+}
+
+export type Snapshots = Record<number, Snapshot>;
+
+export interface WatchGroup {
+  id: number;
+  name: string;
+  category: string;
+  userIds: number[];
+}
+
+export interface WatchData {
+  userIds: number[];
+  snapshots: SnapshotsData;
```

Comment:
Where did you see "heartbeat snapshots"? Did you mean `HeartbeatDetail`?

A snapshot here means strictly a snapshot of _a session_, so there's no need for `sessionSnapshot`. If there's a confusion because there's a "heartbeat snapshot", we can change that one instead. But I didn't recall calling making a snapshot of a heartbeat(?)

`watch` and `pulse` of `LiveMonitoringChannel` both send the same thing: snapshot. Just that `pulse` might omit some constants to make the broadcasts slimmer. In PulseGrid, "session" means just like what it represents, but _the data that's shown and shared to update the sessions_, they are snapshots. Strictly, snapshot represents the _latest_ state of a session. By this definition, since a sessions have many heartbeats, the attributes of a snapshot and heartbeat may be similar, but I didn't want to extend them or make them share anything because they are used in different places. I need the type safety such that they won't accidentally satisfy each other.

Reply:
I didnt see it but the naming of snapshot gave me the impression that it could be heartbeat snapshot because we are only showing the data of the latest heartbeat. Only after reading the rest of the code that I realised it is a snapshot of the session

Reply:
I made sure that in the client code, there's no use of the namespace `session` other than the component names, which are correct (sessions make up the grid), to avoid confusion, because the client doesn't manipulate sessions, only the server. Instead, we have many `snapshot`s, and `heartbeat`s.

I know it isn't perfect, but I think this is the best that I can come up with. Naming things is one of the hardest thing in programming. I spent hours changing names and variables on this project 😂